sympy/polys/polytools.py

```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
+            A non-negative lower bound which will be positive if all the roots
+            are distinct, and rational if the coefficients of the polynomial
+            are rational.
+
+        Raises
+        ======
+
+        DomainError
+            If the domain of the polynomial is not a subring of the complex
+            numbers.
+        MultivariatePolynomialError
+            If the polynomial is not univariate.
+
+        """
+        dom = f.get_domain()
+        if dom.is_FiniteField or not dom.is_Numerical:
+            raise DomainError("Cannot compute root bounds over %s" % dom)
+        if f.is_multivariate:
+            raise MultivariatePolynomialError(
+                "Must be a univariate polynomial")
+        if f.is_monomial:
+            return Integer(0)
+
+        x = f.gen
+        y = Dummy('y')
+        g = f.subs(x, x + y)
+        h = Poly(resultant(f, g, gens=[x, y]), domain=f.domain)
+        # If f has roots ri, the roots of h are all the ri - rj.
+
+        if h.is_monomial:
+            # In this case all roots of f are the same.
+            return Integer(0)
+
+        lb, _ = h.root_bounds()
+
+        # Since h is an even function, can improve bound with square root.
+        if lb.is_rational:
+            # Preserve rational bound
+            return Rational(isqrt(lb.p*lb.q) + 1, lb.q)
+        else:
+            from sympy.functions.elementary.miscellaneous import sqrt
+            return sqrt(lb)
+
+    def root_comparison_tools(f):
+        """
+        Produce a threshold ``eps`` and custom evalf function ``evalf_`` such
+        that two roots ``a``, ``b`` of this polynomial must be equal if
+        ``abs(evalf_(a) - evalf_(b)) < eps``.
+
+        Examples
+        ========
+
+        >>> from sympy import Poly, cyclotomic_poly, exp, I, pi
+        >>> f = Poly(cyclotomic_poly(5))
+        >>> eps, evalf_ = f.root_comparison_tools()
+        >>> print(eps)
+        2/7
+        >>> r0 = exp(2*I*pi/5)
+        >>> R0 = evalf_(r0)
+        >>> print(R0)  # doctest: +SKIP
+        0.309 + 0.9511*I
+        >>> check = lambda r: abs(evalf_(r) - R0) < eps
+        >>> indices = [i for i, r in enumerate(f.all_roots()) if check(r)]
+        >>> print(indices)
+        [3]
+
+        Raises
+        ======
+
+        DomainError
+            If the domain of the polynomial is not a subring of the complex
+            numbers.
+        MultivariatePolynomialError
+            If the polynomial is not univariate.
+
+        """
+        delta = f.root_separation_lower_bound()
+        eps = delta / 3
+
+        # Pick n large enough to guarantee 10^n > 1/eps.
+        L = fastlog10_for_expr
+        n = L(1/eps)
+
+        # Make a custom evalf function that guarantees rounding error < 10^(-n).
+        # This means we want more than n places to the right of the decimal
+        # point. `L(x)` gets us past any places to the left of the point.
+        evalf_ = lambda x: x.evalf(L(x) + n + 1)
```

Comment:
Is this the reason for needing fastlog10?

Maybe `dps_to_prec` could be used here to work in binary precision.

Reply:
Yes, this was part of the reason. Also I wanted a function that would accept any `Expr`, instead of the `MPF_TUP` that's required by `fastlog`.

Maybe, if we try to work with domain elements instead of `Expr`, as you discuss above, this could be reconsidered.

Reply:
The `evalf` function in the `sympy.core.evalf` module takes a binary precision but otherwise does more or less the same as the `Expr.evalf` method.

The mpf tuple is easy to obtain:
```python
In [1]: pi.evalf(1)._mpf_
Out[1]: (0, mpz(25), -3, 5)
```